odl/space/base_tensors.py

```diff
@@ -0,0 +1,1031 @@
+            The default ``None`` is equivalent to ``self.order``.
+
+        Returns
+        -------
+        newspace : `TensorSpace`
+            Version of this space with given data type.
+        """
+        if dtype is None:
+            # Need to filter this out since Numpy iterprets it as 'float'
+            raise ValueError('`None` is not a valid data type')
+
+        if order is None:
+            order = self.order
+
+        dtype = np.dtype(dtype)
+        if dtype == self.dtype and order == self.order:
+            return self
+
+        if is_numeric_dtype(self.dtype):
+            # Caching for real and complex versions (exact dtype mappings)
+            if dtype == self.__real_dtype and order == self.order:
+                if self.__real_space is None:
+                    self.__real_space = self._astype(dtype, self.order)
+                return self.__real_space
+            elif dtype == self.__complex_dtype and order == self.order:
+                if self.__complex_space is None:
+                    self.__complex_space = self._astype(dtype, self.order)
+                return self.__complex_space
+            else:
+                return self._astype(dtype, order)
+        else:
+            return self._astype(dtype, order)
+
+    @property
+    def order(self):
+        """Guaranteed data storage order in this space.
+
+        This is one of ``('A', 'C', 'F')``, where ``'A'`` means "any",
+        i.e. "no guarantee".
+        """
+        return self.__order
+
+    @property
+    def default_order(self):
+        """Storage order for new elements in this space.
+
+        This is identical to `order` except for ``self.order == 'A'``,
+        where ``'C'`` is returned.
+        """
+        return 'C' if self.order == 'A' else self.order
+
+    @property
+    def size(self):
+        """Total number of entries in an element of this space."""
+        return 0 if self.shape == () else int(np.prod(self.shape))
+
+    @property
+    def ndim(self):
+        """Number of axes (=dimensions) of this space, also called "rank"."""
+        return len(self.shape)
+
+    def __len__(self):
+        """Number of tensor entries along the first axis."""
+        return int(self.shape[0])
+
+    @property
+    def itemsize(self):
+        """Size in bytes of one entry in an element of this space."""
+        return int(self.dtype.itemsize)
+
+    @property
+    def nbytes(self):
+        """Total number of bytes used by an element of this space in memory."""
+        return self.size * self.itemsize
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
+
+        This base version of indexing does not propagate weightings.
+        Subclasses need to override this method if weighting propagation
+        is desired.
+
+        Examples
+        --------
+        Integers and slices index the first axis:
+
+        >>> space = odl.rn((2, 3, 4))
+        >>> space[1]
```

Comment:
Yeah, this whole thing gives me some headache. The only reason I can currently see to keep it implemented is compatibility with `ProductSpace`.
Other than that, I really have a hard time thinking of `rn((2, 3))` as a stack of 2 `rn(3)`'s. Indexing something whose constituents are indistinguishable from each other really adds very little value. That is, `space[0] == space[1]` is **always** true. What's the point then?

And the argument that `space[indices]` should be the same as `element[indices].space` doesn't really convince me either. It just means lots of extra code for little gain, since we index `element` in a lazy way anyway. So then one could go directly with `element[indices].space`.

Another reason for removing `__getitem__` on `TensorSpace` is that its big cousin `DiscreteLp` will not implement it since we have a huge ambiguity regarding **what** we want to index. For vector-valued functions, are we after the **input** or **output** dimensions?

I'm happy to hear other arguments. Currently I'm leaning towards dumping this method since "In the face of ambiguity, refuse the temptation to guess." (Zen of Python)

Reply:
Just for the record: simply deleting `__getitem__` of both `TensorSpace` and `NumpyTensorSpace` makes nothing fail except the unit tests designed to test exactly those methods. That means we don't need them.

Reply:
> Yeah, this whole thing gives me some headache. The only reason I can currently see to keep it implemented is compatibility with ProductSpace.

And also there, it's a new feature that may be worth a discussion.

Reply:
I'd move this code to a gist or something and open an issue where we can discuss the future of them. Leave it out of this PR